File: Attack_on_Chaiwala.py
import pygame
import random
import sys

# Initialize Pygame
pygame.init()

# Screen settings
CELL_SIZE = 40
GRID_WIDTH = 20
GRID_HEIGHT = 20
SCREEN_WIDTH = CELL_SIZE * GRID_WIDTH
SCREEN_HEIGHT = CELL_SIZE * GRID_HEIGHT
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
pygame.display.set_caption("Attack on Chaiwala")

# Colors
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)

# Load images
doraemon_img = pygame.transform.scale(pygame.image.load("apa.jpg"), (CELL_SIZE, CELL_SIZE))
dorayaki_img = pygame.transform.scale(pygame.image.load("chaiwala.jpg"), (CELL_SIZE, CELL_SIZE))
rocket_img = pygame.transform.scale(pygame.image.load("trash.jpg"), (CELL_SIZE, CELL_SIZE))
# Load wood texture for background
wood_texture = pygame.transform.scale(pygame.image.load("shojon.jpg"), (SCREEN_WIDTH, SCREEN_HEIGHT))

# Clock
clock = pygame.time.Clock()



# Game variables
def reset_game():
    global doraemon_pos, rockets, direction, dorayaki_pos, score, game_over, paused, speed
    doraemon_pos = [GRID_WIDTH // 2, GRID_HEIGHT // 2]
    rockets = []
    direction = (0, 0)
    dorayaki_pos = [random.randint(0, GRID_WIDTH - 1), random.randint(0, GRID_HEIGHT - 1)]
    score = 0
    game_over = False
    paused = False
    speed = 10  # initial speed (frames per second)

reset_game()

font = pygame.font.SysFont(None, 36)

# Function to draw grid
def draw_grid():
    for x in range(0, SCREEN_WIDTH, CELL_SIZE):
        pygame.draw.line(screen, (200, 200, 200), (x, 0), (x, SCREEN_HEIGHT))
    for y in range(0, SCREEN_HEIGHT, CELL_SIZE):
        pygame.draw.line(screen, (200, 200, 200), (0, y), (SCREEN_WIDTH, y))

# Function to draw scoreboard
def draw_score():
    text = font.render(f"Score: {score}", True, BLACK)
    screen.blit(text, (10, 10))

# Main game loop
while True:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.quit()
            sys.exit()
        elif event.type == pygame.KEYDOWN:
            # Arrow keys and WASD controls
            if not paused and not game_over:
                if (event.key == pygame.K_UP or event.key == pygame.K_w) and direction != (0, 1):
                    direction = (0, -1)
                elif (event.key == pygame.K_DOWN or event.key == pygame.K_s) and direction != (0, -1):
                    direction = (0, 1)
                elif (event.key == pygame.K_LEFT or event.key == pygame.K_a) and direction != (1, 0):
                    direction = (-1, 0)
                elif (event.key == pygame.K_RIGHT or event.key == pygame.K_d) and direction != (-1, 0):
                    direction = (1, 0)
            # Pause/unpause with E
            if event.key == pygame.K_e:
                if not game_over:
                    paused = not paused
            # Restart with R
            if event.key == pygame.K_r:
                reset_game()

    if not paused and not game_over and direction != (0, 0):
        prev_head = list(doraemon_pos)
        # Move Doraemon
        doraemon_pos[0] += direction[0]
        doraemon_pos[1] += direction[1]
        # Wrap around screen edges
        doraemon_pos[0] = doraemon_pos[0] % GRID_WIDTH
        doraemon_pos[1] = doraemon_pos[1] % GRID_HEIGHT

        # Insert previous head at the start of the tail
        if rockets:
            rockets = [prev_head] + rockets

        # Check collisions with dorayaki
        if doraemon_pos == dorayaki_pos:
            # Grow: keep the new tail segment
            dorayaki_pos = [random.randint(0, GRID_WIDTH - 1), random.randint(0, GRID_HEIGHT - 1)]
            score += 1
            # Increase speed moderately after each bunch of dorayaki
            import math
            bump = int(math.log2(score + 1)) if score > 0 else 0
            speed = 10 + bump * 2  # base 10, +2 fps for each power of 2 dorayaki
        else:
            # Move: remove the last tail segment if not eating
            if rockets:
                rockets.pop()

        # If first dorayaki eaten, start tail
        if not rockets and score > 0:
            rockets = [prev_head]

        # Check self collision (rockets)
        if doraemon_pos in rockets:
            game_over = True

    # Drawing
    screen.blit(wood_texture, (0, 0))
    draw_grid()
    if not game_over:
        screen.blit(doraemon_img, (doraemon_pos[0] * CELL_SIZE, doraemon_pos[1] * CELL_SIZE))
        for rocket in rockets:
            screen.blit(rocket_img, (rocket[0] * CELL_SIZE, rocket[1] * CELL_SIZE))
        # If dorayaki overlaps with Doraemon or any rocket, draw a black spot
        if dorayaki_pos == doraemon_pos or dorayaki_pos in rockets:
            center_x = dorayaki_pos[0] * CELL_SIZE + CELL_SIZE // 2
            center_y = dorayaki_pos[1] * CELL_SIZE + CELL_SIZE // 2
            pygame.draw.circle(screen, BLACK, (center_x, center_y), CELL_SIZE // 3)
        else:
            screen.blit(dorayaki_img, (dorayaki_pos[0] * CELL_SIZE, dorayaki_pos[1] * CELL_SIZE))
        draw_score()
        if paused:
            pause_font = pygame.font.SysFont(None, 72)
            pause_text = pause_font.render("Paused", True, (0, 0, 200))
            screen.blit(pause_text, (SCREEN_WIDTH//2 - pause_text.get_width()//2, SCREEN_HEIGHT//2 - 40))
    else:
        # Game over screen
        game_over_font = pygame.font.SysFont(None, 72)
        score_font = pygame.font.SysFont(None, 48)
        go_text = game_over_font.render("Game Over!", True, (200, 0, 0))
        score_text = score_font.render(f"Score: {score}", True, (0, 0, 0))
        restart_text = score_font.render("Press R to Restart", True, (0, 100, 0))
        screen.blit(go_text, (SCREEN_WIDTH//2 - go_text.get_width()//2, SCREEN_HEIGHT//2 - 80))
        screen.blit(score_text, (SCREEN_WIDTH//2 - score_text.get_width()//2, SCREEN_HEIGHT//2))
        screen.blit(restart_text, (SCREEN_WIDTH//2 - restart_text.get_width()//2, SCREEN_HEIGHT//2 + 60))

    pygame.display.flip()
    clock.tick(speed)

File: attack_on_chaiwala.js
// Attack on Chaiwala - JavaScript version using HTML5 Canvas

// Place this in an HTML file with a <canvas id="gameCanvas"></canvas> element

const CELL_SIZE = 30;
const GRID_WIDTH = 50;
const GRID_HEIGHT = 25;
const SCREEN_WIDTH = CELL_SIZE * GRID_WIDTH;
const SCREEN_HEIGHT = CELL_SIZE * GRID_HEIGHT;

const canvas = document.getElementById('gameCanvas');
canvas.width = SCREEN_WIDTH;
canvas.height = SCREEN_HEIGHT;
const ctx = canvas.getContext('2d');

// Load images
const apa_img = new Image(); apa_img.src = 'apa.jpg';
const chaiwala_img = new Image(); chaiwala_img.src = 'chaiwala.jpg';
const trash_img = new Image(); trash_img.src = 'trash.jpg';
const shojon_img = new Image(); shojon_img.src = 'shojon.jpg';

// Game variables
let apa_pos, trashes, direction, chaiwala_pos, score, game_over, paused, speed;

function resetGame() {
    apa_pos = [Math.floor(GRID_WIDTH / 2), Math.floor(GRID_HEIGHT / 2)];
    trashes = [];
    direction = [0, 0];
    chaiwala_pos = [randInt(0, GRID_WIDTH - 1), randInt(0, GRID_HEIGHT - 1)];
    score = 0;
    game_over = false;
    paused = false;
    speed = 10;
}

function randInt(a, b) {
    return Math.floor(Math.random() * (b - a + 1)) + a;
}

function drawGrid() {
    ctx.strokeStyle = '#CCCCCC';
    for (let x = 0; x <= SCREEN_WIDTH; x += CELL_SIZE) {
        ctx.beginPath(); ctx.moveTo(x, 0); ctx.lineTo(x, SCREEN_HEIGHT); ctx.stroke();
    }
    for (let y = 0; y <= SCREEN_HEIGHT; y += CELL_SIZE) {
        ctx.beginPath(); ctx.moveTo(0, y); ctx.lineTo(SCREEN_WIDTH, y); ctx.stroke();
    }
}

function drawScore() {
    ctx.font = 'bold 28px Arial';
    // Shadow
    ctx.fillStyle = 'black';
    ctx.fillText(`Score: ${score}`, 13, 33);
    // Main text
    ctx.fillStyle = 'yellow';
    ctx.fillText(`Score: ${score}`, 10, 30);
}

function drawPaused() {
    ctx.font = 'bold 72px Arial';
    ctx.fillStyle = 'black';
    ctx.fillText('Paused', SCREEN_WIDTH/2 - 117, SCREEN_HEIGHT/2 - 37);
    ctx.fillStyle = 'cyan';
    ctx.fillText('Paused', SCREEN_WIDTH/2 - 120, SCREEN_HEIGHT/2 - 40);
}

function drawGameOver() {
    ctx.font = 'bold 72px Arial';
    ctx.fillStyle = 'black';
    ctx.fillText('Game Over!', SCREEN_WIDTH/2 - 177, SCREEN_HEIGHT/2 - 77);
    ctx.fillStyle = 'red';
    ctx.fillText('Game Over!', SCREEN_WIDTH/2 - 180, SCREEN_HEIGHT/2 - 80);
    ctx.font = 'bold 48px Arial';
    ctx.fillStyle = 'black';
    ctx.fillText(`Score: ${score}`, SCREEN_WIDTH/2 - 78, SCREEN_HEIGHT/2 + 3);
    ctx.fillStyle = 'yellow';
    ctx.fillText(`Score: ${score}`, SCREEN_WIDTH/2 - 80, SCREEN_HEIGHT/2);
    ctx.fillStyle = 'black';
    ctx.fillText('Press R to Restart', SCREEN_WIDTH/2 - 138, SCREEN_HEIGHT/2 + 63);
    ctx.fillStyle = 'lime';
    ctx.fillText('Press R to Restart', SCREEN_WIDTH/2 - 140, SCREEN_HEIGHT/2 + 60);
}

function draw() {
    // Draw background
    ctx.drawImage(shojon_img, 0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);
    drawGrid();
    if (!game_over) {
        ctx.drawImage(apa_img, apa_pos[0] * CELL_SIZE, apa_pos[1] * CELL_SIZE, CELL_SIZE, CELL_SIZE);
        for (let trash of trashes) {
            ctx.drawImage(trash_img, trash[0] * CELL_SIZE, trash[1] * CELL_SIZE, CELL_SIZE, CELL_SIZE);
        }
        // If chaiwala overlaps with Apa or any trash, draw a black spot
        let overlap = (chaiwala_pos[0] === apa_pos[0] && chaiwala_pos[1] === apa_pos[1]) || trashes.some(t => t[0] === chaiwala_pos[0] && t[1] === chaiwala_pos[1]);
        if (overlap) {
            ctx.beginPath();
            ctx.arc(chaiwala_pos[0] * CELL_SIZE + CELL_SIZE/2, chaiwala_pos[1] * CELL_SIZE + CELL_SIZE/2, CELL_SIZE/3, 0, 2*Math.PI);
            ctx.fillStyle = 'black';
            ctx.fill();
        } else {
            ctx.drawImage(chaiwala_img, chaiwala_pos[0] * CELL_SIZE, chaiwala_pos[1] * CELL_SIZE, CELL_SIZE, CELL_SIZE);
        }
        drawScore();
        if (paused) drawPaused();
    } else {
        drawGameOver();
    }
}

let lastMove = 0;
function gameLoop(ts) {
    if (!lastMove) lastMove = ts;
    let interval = 1000 / speed;
    if (!paused && !game_over && (direction[0] !== 0 || direction[1] !== 0) && ts - lastMove > interval) {
        lastMove = ts;
        let prev_head = [apa_pos[0], apa_pos[1]];
        apa_pos[0] += direction[0];
        apa_pos[1] += direction[1];
        // Wrap
        apa_pos[0] = (apa_pos[0] + GRID_WIDTH) % GRID_WIDTH;
        apa_pos[1] = (apa_pos[1] + GRID_HEIGHT) % GRID_HEIGHT;
        // Insert previous head at the start of the tail
        if (trashes.length) trashes = [prev_head].concat(trashes);
        // Check collisions with chaiwala
        if (apa_pos[0] === chaiwala_pos[0] && apa_pos[1] === chaiwala_pos[1]) {
            chaiwala_pos = [randInt(0, GRID_WIDTH - 1), randInt(0, GRID_HEIGHT - 1)];
            score++;
            // Increase speed moderately after each bunch of chaiwala
            let bump = score > 0 ? Math.floor(Math.log2(score + 1)) : 0;
            speed = 10 + bump * 2;
        } else {
            // Move: remove the last tail segment if not eating
            if (trashes.length) trashes.pop();
        }
        // If first chaiwala eaten, start tail
        if (trashes.length === 0 && score > 0) trashes = [prev_head];
        // Check self collision
        for (let trash of trashes) {
            if (apa_pos[0] === trash[0] && apa_pos[1] === trash[1]) game_over = true;
        }
    }
    draw();
    requestAnimationFrame(gameLoop);
}

window.addEventListener('keydown', function(e) {
    if (game_over && (e.key === 'r' || e.key === 'R')) { resetGame(); return; }
    if (e.key === 'e' || e.key === 'E') { paused = !paused; return; }
    if (e.key === 'r' || e.key === 'R') { resetGame(); return; }
    if (paused || game_over) return;
    if ((e.key === 'ArrowUp' || e.key === 'w' || e.key === 'W') && !(direction[0] === 0 && direction[1] === 1)) direction = [0, -1];
    else if ((e.key === 'ArrowDown' || e.key === 's' || e.key === 'S') && !(direction[0] === 0 && direction[1] === -1)) direction = [0, 1];
    else if ((e.key === 'ArrowLeft' || e.key === 'a' || e.key === 'A') && !(direction[0] === 1 && direction[1] === 0)) direction = [-1, 0];
    else if ((e.key === 'ArrowRight' || e.key === 'd' || e.key === 'D') && !(direction[0] === -1 && direction[1] === 0)) direction = [1, 0];
});

resetGame();
requestAnimationFrame(gameLoop);


